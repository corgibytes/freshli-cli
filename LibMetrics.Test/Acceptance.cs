using System;
using System.Collections.Generic;
using System.IO;
using ApprovalTests;
using ApprovalTests.Reporters;
using LibMetrics.Languages.Php;
using LibMetrics.Languages.Ruby;
using Xunit;

namespace LibMetrics.Test
{
  [UseReporter(typeof(DiffReporter))]
  public class Acceptance
  {
    [Fact]
    public void RubyGemsWithGitHistory()
    {
      ManifestFinder.Register<RubyBundlerManifestFinder>();
      FileHistoryFinder.Register<GitFileHistoryFinder>();

      var runner = new Runner();

      var rubyFixturePath = Fixtures.Path("ruby", "nokotest");
      var results = runner.Run(rubyFixturePath);

      AssertNokoFixtureRubyGemsResults(results);
    }

    [Fact]
    public void RubyGemsWithHistoryViaGitHub() {
      ManifestFinder.Register<RubyBundlerManifestFinder>();
      FileHistoryFinder.Register<GitFileHistoryFinder>();

      var runner = new Runner();

      var repoUrl =
        "https://github.com/corgibytes/libmetrics-fixture-ruby-nokotest";
      var results = runner.Run(repoUrl);

      AssertNokoFixtureRubyGemsResults(results);
    }

    [Fact]
    public void RubyGemsFeedbinHistoryViaGitHub()
    {
      ManifestFinder.Register<RubyBundlerManifestFinder>();
      FileHistoryFinder.Register<GitFileHistoryFinder>();

      var runner = new Runner();

      var repoUrl = "https://github.com/feedbin/feedbin";
      var results = runner.Run(repoUrl);

      var output = new StringWriter();

      foreach (var resultSet in results) {
        if (resultSet.Date <= new DateTime(2020, 01, 01)) {
          output.WriteLine(
            $"{resultSet.Date.ToString("yyyy/MM/dd")}: " +
            $"{resultSet.LibYear.Total.ToString("F3")}"
          );
        }
      }

      Approvals.Verify(output);
    }

    private static void AssertNokoFixtureRubyGemsResults(IList<MetricsResult> results) {
      // nokotest contains a git repository with the following history:
      // Gemfile v1 created on 2017/01/01 references nokogiri 1.7.0
      // Gemfile v2 created on 2018/01/01 references nokogiri 1.8.1
      // Gemfile v3 created on 2019/01/01 references nokogiri 1.9.1

      var output = new StringWriter();

      foreach (var resultSet in results) {
          if (resultSet.Date <= new DateTime(2020, 01, 01)) {
              output.WriteLine(
                  $"{resultSet.Date.ToString("yyyy/MM/dd")}: " +
                  $"{resultSet.LibYear.Total.ToString("F3")}"
              );
          }
      }

      Approvals.Verify(output);
    }

    [Fact]
    public void ComposerWithoutGitHistory()
    {
      ManifestFinder.Register<PhpComposerManifestFinder>();

      var runner = new Runner();

      var phpFixturePath = Fixtures.Path("php", "large");
      var results = runner.Run(
        phpFixturePath,
        asOf: new DateTime(2020, 01, 01));

      var output = new StringWriter();

      foreach (var resultSet in results)
      {
        output.WriteLine($"{resultSet.Date.ToString("yyyy/MM/dd")}: " +
                         $"{resultSet.LibYear.Total.ToString("F3")}");
      }

      var expected = new StringWriter();
      expected.WriteLine("2020/01/01: 6.126");

      Assert.Equal(expected.ToString(), output.ToString());
    }

    [Fact]
    public void DrupalComposerWithoutGitHistory()
    {
      ManifestFinder.Register<PhpComposerManifestFinder>();

      var runner = new Runner();

      var phpFixturePath = Fixtures.Path("php", "drupal");
      var results = runner.Run(
        phpFixturePath,
        asOf: new DateTime(2020, 01, 01));

      var output = new StringWriter();

      foreach (var resultSet in results)
      {
        output.WriteLine($"{resultSet.Date.ToString("yyyy/MM/dd")}: " +
                         $"{resultSet.LibYear.Total.ToString("F3")}");
      }

      var expected = new StringWriter();
      expected.WriteLine("2020/01/01: 3.214");

      Assert.Equal(expected.ToString(), output.ToString());
    }
  }
}
