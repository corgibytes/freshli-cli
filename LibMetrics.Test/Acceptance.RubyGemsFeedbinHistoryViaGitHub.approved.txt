2013/09/01: 7.444
2013/10/01: 4.649
2013/11/01: 8.389
2013/12/01: 20.879
2014/01/01: 11.762
2014/02/01: 10.986
2014/03/01: 17.814
2014/04/01: 10.940
2014/05/01: 21.677
2014/06/01: 16.858
2014/07/01: 38.778
2014/08/01: 26.989
2014/09/01: 32.118
2014/10/01: 36.019
2014/11/01: 41.945
2014/12/01: 25.701
2015/01/01: 24.353
2015/02/01: 38.819
2015/03/01: 48.411
2015/04/01: 56.499
2015/05/01: 65.767
2015/06/01: 68.696
2015/07/01: 18.060
2015/08/01: 25.814
2015/09/01: 13.973
2015/10/01: 22.030
2015/11/01: 30.789
2015/12/01: 18.279
2016/01/01: 25.304
2016/02/01: 20.351
2016/03/01: 32.658
2016/04/01: 26.863
2016/05/01: 37.296
2016/06/01: 23.279
2016/07/01: 22.022
2016/08/01: 34.249
2016/09/01: 17.077
2016/10/01: 22.512
2016/11/01: 26.742
2016/12/01: 43.586
2017/01/01: 25.370
2017/02/01: 37.532
2017/03/01: 51.334
2017/04/01: 61.827
2017/05/01: 72.652
2017/06/01: 11.934
2017/07/01: 21.200
2017/08/01: 16.279
2017/09/01: 15.085
2017/10/01: 24.600
2017/11/01: 29.740
2017/12/01: 25.940
2018/01/01: 28.885
2018/02/01: 47.005
2018/03/01: 33.636
2018/04/01: 45.693
2018/05/01: 54.318
2018/06/01: 61.781
2018/07/01: 36.395
2018/08/01: 24.477
2018/09/01: 28.427
2018/10/01: 32.351
2018/11/01: 43.534
2018/12/01: 61.312
2019/01/01: 38.600
2019/02/01: 43.090
2019/03/01: 33.748
2019/04/01: 48.912
2019/05/01: 61.841
2019/06/01: 39.671
2019/07/01: 37.249
2019/08/01: 48.112
2019/09/01: 44.671
2019/10/01: 55.962
2019/11/01: 72.252
2019/12/01: 75.430
2020/01/01: 34.868
